ktor-client/ktor-client-cio/common/src/io/ktor/client/engine/cio/utils.kt

```diff
@@ -138,6 +138,55 @@ internal suspend fun readResponse(
     }
 }
 
+internal suspend fun startTunnel(
+    request: HttpRequestData,
+    output: ByteWriteChannel,
+    input: ByteReadChannel
+) {
+    val builder = RequestResponseBuilder()
+
+    try {
+        val urlString = request.url.toString()
+
+        builder.requestLine(HttpMethod("CONNECT"), urlString, HttpProtocolVersion.HTTP_1_1.toString())
+        // this will only add the port to the host header if the port is non-standard for the protocol
+        val host = if (request.url.protocol.defaultPort == request.url.port) {
+            request.url.host
+        } else {
+            request.url.hostWithPort
+        }
+
+        builder.headerLine(HttpHeaders.Host, host)
+        builder.headerLine("Proxy-Connection", "Keep-Alive") // For HTTP/1.0 proxies like Squid.
+        request.headers[HttpHeaders.UserAgent]?.let {
+            builder.headerLine(HttpHeaders.UserAgent, it)
```

Comment:
I wonder, do we need to pass it? It could be considered as a security issue as it deanonymizes user. Sorry for the old link, can't quickly find the actual document: https://www.eff.org/deeplinks/2010/01/tracking-by-user-agent